### ISSUE TITLE
Range-versioned dependencies are not correctly found with overridden repos configuration

### ISSUE THREAD

**Issue by dlecan** [sbt/sbt#752]
## steps

@joescii reproduced it on https://github.com/joescii/sbt-reproduced/tree/issue-752
## original report

Depending on a Maven artifact, which depends transitively on a range-versioned dependency, doesn't work if repos configuration is overridden.

SBT launcher: 0.12.3

With default repos configuration:

```
[info] Resolving org.avaje.ebeanorm#avaje-ebeanorm-agent;[3.1,4) ...
...
[info] downloading http://repo1.maven.org/maven2/org/avaje/ebeanorm/avaje-ebeanorm-agent/3.2.1/avaje-ebeanorm-agent-3.2.1.jar ...
[info]  [SUCCESSFUL ] org.avaje.ebeanorm#avaje-ebeanorm-agent;3.2.1!avaje-ebeanorm-agent.jar (7ms)
```

With overridden repos configuration:

```
[info] Resolving org.avaje.ebeanorm#avaje-ebeanorm-agent;[3.1,4) ...
[warn]  module not found: org.avaje.ebeanorm#avaje-ebeanorm-agent;[3.1,4)
[warn] ==== local: tried
[warn]   /home/jenkins/.ivy2/local/org.avaje.ebeanorm/avaje-ebeanorm-agent/[revision]/ivys/ivy.xml
[warn] ==== maven-central-cloudbees-mirror: tried
[warn]   http://repo.cloudbees.com/content/groups/public/org/avaje/ebeanorm/avaje-ebeanorm-agent/[revision]/avaje-ebeanorm-agent-[revision].pom
[warn] ==== sonatype-snapshots: tried
[warn]   https://oss.sonatype.org/content/repositories/snapshots/org/avaje/ebeanorm/avaje-ebeanorm-agent/[revision]/avaje-ebeanorm-agent-[revision].pom
```

Notice the `[revision]` in URLs, which is not substituted by right revision.

**Comment by harrah** [comment]
Can you provide a test case?

**Comment by aindlq** [comment]
I am facing the same issue, I was going to setup internal repository and use it as proxy according to documentation: http://www.scala-sbt.org/release/docs/Detailed-Topics/Proxy-Repositories.html
but it cant substitute correct dependency version for org.avaje.ebeanorm#avaje-ebeanorm-agent;[3.1,4)

**Comment by opyate** [comment]
@hsn10 Has #752 and #779 been closed as duplicates of one another?

Please open this one again, as it came first.

**Comment by harrah** [comment]
This one is not closed, although it does not have a reproducible test case and so it perhaps should be closed.

**Comment by cambler-godaddy** [comment]
Is this being investigated at all? I'm having the same issue.

**Comment by jsuereth** [comment]
@cambler If you get me a test case reproduction, I'll see if I can fix it.

**Comment by joescii** [comment]
I've hit this issue as well, Josh.  I believe I could stand up an Archiva maven server in our AWS with a project that reproduces it.  Let me know when you think you'll have time to check this out, and I'll see what I can do.

**Comment by freewind** [comment]
The same error I've hit, but hard for me to provide a test case. Just reference the issue I've reports to another project [repox#9](https://github.com/Centaur/repox/issues/9)

**Comment by joescii** [comment]
Yo, @jsuereth... It's easily reproducible [here](https://github.com/joescii/sbt-reproduced/tree/issue-752).

**Comment by freewind** [comment]
@joescii,thanks! Hope it get fixed soon

**Comment by eed3si9n** [comment]
@joescii Thanks for the repro case.
In both yours and the original case, the dynamic revision is specified incorrectly according to Ivy's [dependency](http://ant.apache.org/ivy/history/2.3.0/ivyfile/dependency.html), which different from Maven's [Version Range Specification](http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html).
- Resolving org.avaje.ebeanorm#avaje-ebeanorm-agent;[3.1,4)
- Resolving rubygems#middleman-core;[3.3.7,3.3.7.0.0.0.0.1)

In other words, you need `[` to mean less-than instead of `)`.
<s>Maybe we should issue a warning about this, but this is not a bug as far as sbt is concerned.</s>

**Update**: sbt seems to be ok with Maven style version range.

**Comment by joescii** [comment]
I see... So the repos in the overrides must be ivy style whereas a resolver can be either?

Joe

**Comment by eed3si9n** [comment]
@joescii I've been looking into this more.
As you suggested, sbt does support Maven repository on its resolvers, so it _should_ handle Maven's version semantics.

There are some puzzling things, however. I tried running your repro case locally, and my failure is different from what's on your README:

```
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: rubygems#middleman;3.3.7: rubygems-release: unable to get resource for rubygems#middleman;3.3.7: res=http://rubygems-proxy.torquebox.org/releases/: java.text.ParseException: null in file:/var/folders/xxx/xxx/T/ivy6973646341199185799xml
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn] 
[warn]  Note: Unresolved dependencies path:
[warn]    rubygems:middleman:3.3.7 (/xxx/xxx/sbt-reproduced/build.sbt#L12-15)
[warn]      +- com.joescii:sbt-reproduced-752_2.11:0.0.1-SNAPSHOT
sbt.ResolveException: unresolved dependency: rubygems#middleman;3.3.7: rubygems-release: unable to get resource for rubygems#middleman;3.3.7: res=http://rubygems-proxy.torquebox.org/releases/: java.text.ParseException: null in file:/var/folders/xxx/xxx/T/ivy6973646341199185799xml
```

Next I went looking for other examples of Maven version range.

```
# if you have silver searcher
$ ag '<version>\[' ~/.ivy2/cache | less 
# if not, the slow version
$ grep -R '<version>\[' ~/.ivy2/cache
```

One I found is:

``` scala
libraryDependencies += "com.amazonaws" % "aws-java-sdk" % "1.2.6"
```

[pom](http://central.maven.org/maven2/com/amazonaws/aws-java-sdk/1.2.6/aws-java-sdk-1.2.6.pom) contains:

``` xml
<dependency>
  <groupId>commons-logging</groupId>
  <artifactId>commons-logging</artifactId>
  <version>[1.1, 2.0)</version>
</dependency>
<dependency>
  <groupId>org.apache.httpcomponents</groupId>
  <artifactId>httpclient</artifactId>
  <version>[4.1, 5.0)</version>
</dependency>
```

When I run this, sbt seems ok with the versions:

```
[info] Resolving commons-logging#commons-logging;[1.1, 2.0) ...
[info] Resolving commons-logging#commons-logging;[1.1, 2.0) ...
[info] Resolving org.apache.httpcomponents#httpclient;[4.1, 5.0) ...
[info] Resolving org.apache.httpcomponents#httpclient;[4.1, 5.0) ...
...
[info] Done updating.
```

Then it finally dawned on me 5:00 in the morning that the problem could be in your [`repositories`](https://github.com/joescii/sbt-reproduced/blob/a1116fe3fb3fab46a84d8041a18363d2d349a9f4/sbt/conf/repositories) file:

```
[repositories]
  local
  typesafe-releases: http://repo.typesafe.com/typesafe/ivy-releases/, [organization]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]
  sbt-plugin-releases: http://repo.scala-sbt.org/scalasbt/sbt-plugin-releases/, [organization]/[module]/(scala_[scalaVersion]/)(sbt_[sbtVersion]/)[revision]/[type]s/[artifact](-[classifier]).[ext]
  rubygems-release: http://rubygems-proxy.torquebox.org/releases/,
  maven-central-releases: http://repo1.maven.org/maven2
```

Sure enough you have an extra comma after the URL without specifying the repository layout.
I removed the comma, and sure enough, the middleman started resolving:

```
[info] Resolving rubygems#middleman;3.3.7 ...
[info] Resolving rubygems#middleman;3.3.7 ...
[info] Resolving rubygems#middleman-core;[3.3.7,3.3.7.0.0.0.0.1) ...
[info] Resolving rubygems#middleman-core;[3.3.7,3.3.7.0.0.0.0.1) ...
[info] Resolving rubygems#bundler;[1.1,1.99999] ...
[info] Resolving rubygems#bundler;[1.1,1.99999] ...
```

**Comment by joescii** [comment]
That's funny because at some point I added the comma to get it to work at all.  Maybe I had some other problem and I never went back to undo the unneeded comma.  I removed it and it seems to work here.  Now I've got to figure out what I've done wrong with my repos file at work.

Glad this old issue can be closed tho. :+1:

**Comment by greenrd** [comment]
Please reopen this bug. Doing a packet capture with tshark, it seems that the problem is actually that sbt - when a version from a range is requested - literally retrieves a directory listing from the proxy repository, but a directory listing from Nexus only lists the versions that Nexus has currently proxied, not all versions. So a workaround is to manually request a particular version from Nexus, and then it will work thereafter - which explains why this bug has been so hard to reproduce. (Another factor that would make it hard to reproduce is if you didn't delete your ivy cache, you wouldn't see it again after that version had been downloaded.)

**Comment by eed3si9n** [comment]
@greenrd Which sbt version did you use? The screen scraping behavior is an Ivy fall back when it can't find maven-metadata.xml, but until 0.13.9 (#2075) sbt had been falling back every time because we customized the repository pattern.

Whether sbt uses maven-metadata.xml or does the screen scraping, the issue remains that it probably won't find the latest version when it hits the proxy. I personally think that's exactly how it should behave since that's what it found on the repository.

**Comment by greenrd** [comment]
@eed3si9n I'm using sbt 0.13.8 so that makes sense.

So you think that my workaround should be the correct way to deal with this problem? Surely Maven must have a better way of dealing with this issue?

Anyway, I will try using 0.13.9. Maybe Nexus merges the maven-metadata.xml so maybe it will work.

**Comment by greenrd** [comment]
Yes, it works in 0.13.9 - thanks @eed3si9n! I assume that it downloads maven-metadata.xml from the proxied repo.
